src/utils/generateCanonicalUrl.ts

```diff
@@ -0,0 +1,13 @@
+export function generateCanonicalUrl(baseUrl: string, slug?: string) {
+  const cleanBaseUrl = baseUrl.replace(/\/$/, '');
+
+  if (!slug) return `${cleanBaseUrl}/`;
+
+  const sanitizedSlug = slug
+    .replace(/^\/|\/$/g, '')
+    .split('/')
+    .map((segment) => encodeURIComponent(segment))
+    .join('/');
```

Comment:
You can simplify it without regex and this encode URI (this will defeat your purpose as it'll transform chars by unwanted characters).

```suggestion
  const sanitizedSlug = slug
    .trim()
    .split('/')
    .filter(item => item)
    .join('/');
```